src/server/inferUtilityTypes.ts

```diff
@@ -0,0 +1,56 @@
+import type { McpServer, ToolDef } from "./server.js";
+
+/**
+ * Any tool registry shape (includes both widgets and regular tools).
+ * Used as a constraint for type parameters that accept tool registries.
+ */
+export type AnyToolRegistry = Record<string, ToolDef>;
+
+/**
+ * Extract the tool registry type from an McpServer instance.
+ * This includes both widgets (registered via widget()) and regular tools (registered via registerTool()).
+ *
+ * @example
+ * ```ts
+ * type MyTools = InferTools<MyServer>;
+ * // { "search": ToolDef<...>, "calculate": ToolDef<...> }
+ * ```
+ */
+export type InferTools<T> = T extends McpServer<infer W> ? W : never;
+type ExtractTool<T, K extends ToolNames<T>> = InferTools<T>[K];
```

Comment:
Same here and in the rest of the PR

```suggestion
type ExtractTool<ServerType, ToolName extends ToolNames<ServerType>> =
  InferTools<ServerType>[ToolName];
```